src/impression.js

```diff
@@ -47,15 +85,66 @@ export function maybeTrackImpression(win) {
     // avoid duplicate tracking.
     win.location.hash = '';
   }
+
   viewer.whenFirstVisible().then(() => {
-    invoke(win, clickUrl);
+    // TODO(@zhouyx) need test with a real response.
+    const promise = invoke(win, dev().assertString(clickUrl)).then(response => {
+      applyResponse(win, viewer, response);
+    });
+
+    // Timeout invoke promise after 8s and resolve trackImpressionPromise.
+    timerFor(win).timeoutPromise(TIMEOUT_VALUE, promise,
+        'timeout waiting for ad server response').catch(() => {
+        }).then(() => {
```

Comment:
`resolveImpression` is the `resolve` (I hate that that's what everyone calls it, it should be `fulfill`) given to you by the Promise constructor. It takes a single param, and if that param is a Promise itself, it will adopt the param's state (will resolve when the param resolves, with whatever the param resolves to).